thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/cache/DefaultTimeSeriesCache.java

```diff
@@ -0,0 +1,154 @@
+      result = fetchSliceFromSource(slice);
+      insertTimeSeriesIntoCache(result);
+      cacheResponse.mergeSliceIntoRows(result, MergeSliceType.PREPEND);
+    }
+
+    if (cacheResponse.isMissingEndSlice(requestSliceEnd)) {
+      // we add one time granularity to start because the start is inclusive.
+      slice = MetricSlice.from(metricId, cacheResponse.getLastTimestamp() + request.getGroupByTimeGranularity().toMillis(), requestSliceEnd, request.getFilterSet(), request.getGroupByTimeGranularity());
+      result = fetchSliceFromSource(slice);
+      insertTimeSeriesIntoCache(result);
+      cacheResponse.mergeSliceIntoRows(result, MergeSliceType.APPEND);
+    }
+  }
+
+  private ThirdEyeResponse fetchSliceFromSource(MetricSlice slice) throws Exception {
+    TimeSeriesRequestContainer rc = DataFrameUtils.makeTimeSeriesRequestAligned(slice, "ref", this.metricDAO, this.datasetDAO);
+    return this.queryCache.getQueryResult(rc.getRequest());
+  }
+
+  private ThirdEyeResponse buildThirdEyeResponseFromCacheResponse(ThirdEyeCacheResponse cacheResponse) {
+
+    List<String[]> rows = new ArrayList<>();
+    ThirdEyeRequest request = cacheResponse.getCacheRequest().getRequest();
+
+    String dataset = request.getMetricFunctions().get(0).getDataset();
+    DatasetConfigDTO datasetDTO = datasetDAO.findByDataset(dataset);
+    TimeSpec timeSpec = ThirdEyeUtils.getTimeSpecFromDatasetConfig(datasetDTO);
+    DateTimeZone timeZone = DateTimeZone.forID(datasetDTO.getTimezone());
+
+    for (TimeSeriesDataPoint dataPoint : cacheResponse.getRows()) {
+      int timeBucketIndex = TimeRangeUtils.computeBucketIndex(
+          request.getGroupByTimeGranularity(), request.getStartTimeInclusive(), new DateTime(dataPoint.getTimestamp(), timeZone));
+      String dataValue = dataPoint.getDataValue();
+
+      String[] row = new String[2];
+      row[0] = String.valueOf(timeBucketIndex);
+      row[1] = (dataValue == null || dataValue.equals("null")) ? "0" : dataValue;
+
+      rows.add(row);
+    }
+
+    return new RelationalThirdEyeResponse(request, rows, timeSpec);
+  }
+
+  public void insertTimeSeriesIntoCache(ThirdEyeResponse response) {
+
+    // we can consider using fixedThreadPool instead.
+    ExecutorService executor = Executors.newCachedThreadPool();
+
+    // insert points in parallel
+    for (MetricFunction metric : response.getMetricFunctions()) {
+      String metricUrn = MetricEntity.fromMetric(response.getRequest().getFilterSet().asMap(), metric.getMetricId()).getUrn();
+      for (int i = 0; i < response.getNumRowsFor(metric); i++) {
+        Map<String, String> row = response.getRow(metric, i);
+        TimeSeriesDataPoint dp = new TimeSeriesDataPoint(metricUrn, Long.parseLong(row.get(CacheConstants.TIMESTAMP)), metric.getMetricId(), row.get(metric.toString()));
+        executor.execute(() -> this.cacheDAO.insertTimeSeriesDataPoint(dp));
```

Comment:
Maybe consider batch upsert or at least add TODO here to optimize.

Reply:
Currently we are doing parallel insertion, not sure if batch upsert will be faster or not. Will get back to you on this one